core/lib/zksync_core/src/state_keeper/seal_criteria/mod.rs

```diff
@@ -177,6 +177,29 @@ impl IoSealCriteria for TimeoutSealer {
     }
 }
 
+#[derive(Debug, Clone, Copy)]
+pub(super) struct MiniblockMaxPayloadSizeSealer {
+    max_payload_size: usize,
+}
+
+impl MiniblockMaxPayloadSizeSealer {
+    pub fn new(config: &StateKeeperConfig) -> Self {
+        Self {
+            max_payload_size: config.miniblock_max_payload_size,
+        }
+    }
+}
+
+impl IoSealCriteria for MiniblockMaxPayloadSizeSealer {
+    fn should_seal_l1_batch_unconditionally(&mut self, _manager: &UpdatesManager) -> bool {
+        false
+    }
+
+    fn should_seal_miniblock(&mut self, manager: &UpdatesManager) -> bool {
+        manager.miniblock.payload_encoding_size >= self.max_payload_size
```

Comment:
Just to understand better: Are we OK with the situation where the resulting block size significantly exceeds the desired one? You've written about this in the PR description (IMO, it'd make sense to add the corresponding doc comment on `MiniblockMaxPayloadSizeSealer` so that it's not lost), but I'm not entirely sure what prevents a situation when the miniblock size exceeds the consensus threshold.

Reply:
This approach assumes that a miniblock with a single transaction won't exceed the consensus threshold. I'm not sure if `L1`/`ProtocolUpgrade` tx types have a max size limit (beyond EVM constrains), but they cannot be rejected as single transactions anyway. `L2` can be rejected, but have a limit in API level. 

Then, given a single transaction in a miniblock, the problematic case is when it's close but below the state keeper threshold, and then a large transaction is added. For that, there should be a significant gap between the state keeper config, and the actual consensus threshold config, which should also account for additional fields in payload, and potential errors in estimating the encoded payload size. 

A different approach might be to support the exclusion of a secondary transaction in a miniblock, to be rolled over to the next one. This will allow to reduce the gap between the state keeper config and the consensus threshold, but it adds some complexity, still must accept a single transaction, and anyway won't eliminate the need to keep some gap, because it's currently not feasible to build the payload incrementally, and so the size value is just a heuristic.